rfc/composite_table_availability.pseudocode

```diff
@@ -0,0 +1,123 @@
+
+PhysicalTableSchema:
+	getFieldName(Column c)
+
+DataSource:
+	String getName()
+	Map<FieldName, Set<Segment>> raw segment data
+	Map<FieldName, List<Interval>> squashedFromSegments  // optional optimization
+
+	Set<FieldName> getFields() //union of all keys
+	List<Interval> getIntervals(FieldName fieldName)
+	Map<FieldName,List<Interval>> getIntervals(Set<FieldName>) // map slice from squashedMap
+	Set<SegmentId> getSegments(List<Interval>)  // Filter all segments by overlap on time
+
+DataSourceConstraints // Simple bean parameter object
+
+	getGrain
+	Set<FieldName> getMetrics
+	Set<Dimension> getDimensions
+	getApiFilters
+
+	DataSourceConstraints(ApiRequest)
+	DataSourceConstraints (DruidQuery, ApiFilter apifilters)
```

Comment:
There are two use cases.
One is during query planning.
The other is during partial data evaluation during request processing.
The latter could happen AFTER the query has undergone transformations which alter it's field signature.